src/nnet3/nnet-am-decodable-simple.h

```diff
@@ -213,6 +219,10 @@ class NnetDecodableBase {
   // online_ivector_period_ helps us interpret online_ivector_feats_; it's the
   // number of frames the rows of ivector_feats are separated by.
   int32 online_ivector_period_;
+  // In the multiple ivectors case, it is the interval between two consecutive
+  // frames that we want to get ivectors for, so that we can dump multiple
+  // ivectors for a chunk. In the single ivector case, its value is just 0.
```

Comment:
Yiming, I think most of this code is OK now, but I don't think the way you
are working out the input frequency of the iVectors is very elegant.  I
have a suggestion for a different way.

Have a function:

 /// This function works out the interval at which a particular named
 /// input (e.g. input_name == "input" or input_name == "ivector") is
 /// used.  If the input is not used at all this function returns -1;
 /// if it is potentially used on every frame (every "t" value) it returns 1
 /// if it is only used on frame 0 it returns 0; and if it is used only
 /// on multiples of n frames with n > 1, it returns n.
 int32 GetInputInterval(const Nnet &nnet, std::string input_name);

Yiming, I think the easiest way to implement this will be as follows.
Iterate over the network nodes and for each one that has a descriptor (of
types kInput, kDescriptor), print it to a string using Print().  Then parse
the string as a GeneralDescriptor-- this will be easier to write a
recursive function to operate on.

For each descriptor in the network you'll get a prototype/partial
input-interval.
It will be a recursive function, returning values such as -1, 0, 1, 2 and
so on.
Think carefully as you write the recursive function about all the cases.
For example,
for a kOffset descriptor, if the function called on its argument returns 3,
that would normally be considered an interval of 3, but then if the offset
descriptor offsets the 't' value by other than a multiple of 3, it should
return 1 because we're no longer getting the 't' at multiples of a number--
it could be potentially every frame.

1 is the value you return if something unexpected happens such as when the
input is used on a fixed frame not equal to zero.  It's OK on 'hard' cases
if you return 1 even if you 'should' have returned a value >1-- for
instance, if you have two offset descriptors that cancel out each others'
effect.

At the end when you have all the values, in general you'll need to do a GCD
operation, but think about it carefully.

Dan



On Mon, Mar 14, 2016 at 5:04 PM, Yiming Wang <notifications@github.com>
wrote:

> In src/nnet3/nnet-am-decodable-simple.h
> <https://github.com/kaldi-asr/kaldi/pull/468#discussion_r56075629>:
>
> > @@ -213,6 +219,10 @@ class NnetDecodableBase {
> >    // online_ivector_period_ helps us interpret online_ivector_feats_; it's the
> >    // number of frames the rows of ivector_feats are separated by.
> >    int32 online_ivector_period_;
> > +  // In the multiple ivectors case, it is the interval between two consecutive
> > +  // frames that we want to get ivectors for, so that we can dump multiple
> > +  // ivectors for a chunk. In the single ivector case, its value is just 0.
>
> ivector_/online_ivectors_ exists in the original code, so I think they are
> not intended to determine single/multiple ivectors case.
>
> In the config generation script, I use if ivector_interval == 0 to
> determine whether to use ReplaceIndex or Round for ivector. Also in
> nnet3-get-egs, I use the same condition to determine single/multiple
> ivector case.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/kaldi-asr/kaldi/pull/468/files#r56075629>.
>